diffrax/_kl_term.py

```diff
@@ -0,0 +1,175 @@
+import operator
+from typing import Tuple, Callable
+
+import jax
+import jax.numpy as jnp
+import jax.tree_util as jtu
+import equinox as eqx
+from jaxtyping import Array, PyTree
+from ._custom_types import RealScalarLike, Control
+from ._term import (
+    AbstractTerm,
+    ControlTerm,
+    MultiTerm,
+    WeaklyDiagonalControlTerm,
+)
+
+
+def _kl_diagonal(drift: Array, diffusion: Array):
+    """This is the case where diffusion matrix is
+    a diagonal matrix
+    """
+    diffusion = jnp.where(
+        jax.lax.stop_gradient(diffusion) > 1e-7,
+        diffusion,
+        jnp.full_like(diffusion, fill_value=1e-7) * jnp.sign(diffusion),
+    )
+    scale = drift / diffusion
+    return 0.5 * jnp.sum(scale**2)
+
+
+def _kl_full_matrix(drift: Array, diffusion: Array):
+    """General case"""
+    scale = jnp.linalg.pinv(diffusion) @ drift
+    return 0.5 * jnp.sum(scale**2)
+
+def _handle(drift: Array, diffusion: Array):
+    """According to the shape of drift and diffusion,
+    select the right way to compute KL divergence
+    """
+    eqx.error_if(drift, not eqx.is_array(drift), "Only array drifts are supported")
+    eqx.error_if(diffusion, not eqx.is_array(diffusion), "Only array diffusion are supported")
```

Comment:
Note that you need to write `x = eqx.error_if(x, ...)`; otherwise these error checks get DCE'd.